### PR TITLE
Implement do while loops

### PR DESCRIPTION
**Pull request by NeeEoo**

### CODE REVIEW

File: Compiler/FunctionCompiler.py [hunk 1]
```diff
@@ -785,7 +785,30 @@ def compile_while(self): #  while (expression) statement       note - statement
         code += inner_scope_code  # <while> scope code. after this code, pointer points to the next available cell. i.e one after the expression
         code += expression_code  # re-evaluate the expression
         code += "<"  # point to the expression
-        code += "]"  # after <if> scope
+        code += "]"  # after <while> scope
+
+        return code
+
+    def compile_do_while(self):  # do statement while (expression)       note - statement can be scope { }
```
Comment (elikaski) [inline Compiler/FunctionCompiler.py:792]: do statement while (expression) semicolon

File: Compiler/FunctionCompiler.py [hunk 2]
```diff
@@ -785,7 +785,30 @@ def compile_while(self): #  while (expression) statement       note - statement
...
+
+        inner_scope_code = self.compile_statement()
+
+        self.parser.check_current_tokens_are([Token.WHILE, Token.LPAREN])
+        self.parser.advance_token(amount=2)  # point to after LPAREN
+
+        expression_code = self.compile_expression()
+
+        self.parser.check_current_tokens_are([Token.RPAREN, Token.SEMICOLON])
+        self.parser.advance_token()  # point to after RPAREN
```
Comment (elikaski) [inline Compiler/FunctionCompiler.py:804]: self.parser.advance_token(2)  # point to after SEMICOLON

Comment (elikaski) [reply Compiler/FunctionCompiler.py:804]: (otherwise the semicolon is "caught" as empty statement (line 983)

File: Compiler/FunctionCompiler.py [hunk 3]
```diff
@@ -785,7 +785,30 @@ def compile_while(self): #  while (expression) statement       note - statement
...
+
+        code = "[-]+"  # set expression to 1. since do while loops executes the scope code first.
+        code += "["  # go in scope
+        code += inner_scope_code  # <do-while> scope code. after this code, pointer points to the next available cell. i.e one after the expression
+        code += expression_code  # evaluate the expression
```
Comment (elikaski) [inline Compiler/FunctionCompiler.py:809]: after this code, pointer points to the next available cell. i.e one after the expression

File: Compiler/FunctionCompiler.py [hunk 4]
```diff
@@ -785,7 +785,30 @@ def compile_while(self): #  while (expression) statement       note - statement
...
+        self.parser.advance_token(amount=2)  # point to after LPAREN
+
+        expression_code = self.compile_expression()
+
+        self.parser.check_current_tokens_are([Token.RPAREN, Token.SEMICOLON])
+        self.parser.advance_token()  # point to after RPAREN
+
+        code = "[-]+"  # set expression to 1. since do while loops executes the scope code first.
+        code += "["  # go in scope
+        code += inner_scope_code  # <do-while> scope code. after this code, pointer points to the next available cell. i.e one after the expression
```
Comment (elikaski) [inline Compiler/FunctionCompiler.py:808]: <do-while> scope code. after this code, pointer points to the same cell. i.e the expression

File: Compiler/FunctionCompiler.py (hunk 3 above)
Comment (NeeEoo) [reply Compiler/FunctionCompiler.py:809]: `# evaluate the expression, after this code, the pointer is pointing to the next cell`
Is this better

Comment (elikaski) [reply Compiler/FunctionCompiler.py:809]: sure

**Comment by elikaski** [comment]
Great work! Thank you

**Pull request merged**